src/ff1.rs

```diff
@@ -146,6 +146,95 @@ impl RadixOps for PowerTwoRadix {
     }
 }
 
+/// A numeral string with radix 2.
+pub struct BinaryNumeralString(Vec<u8>);
+
+impl BinaryNumeralString {
+    /// Creates a BinaryNumeralString from a byte slice, with each byte
+    /// interpreted in little-endian bit order.
+    pub fn from_bytes_le(s: &[u8]) -> Self {
+        let mut data = Vec::with_capacity(s.len() * 8);
+        for n in s {
+            let mut tmp = *n;
+            for _ in 0..8 {
+                data.push(tmp & 1);
+                tmp >>= 1;
+            }
+        }
+        BinaryNumeralString(data)
+    }
+
+    /// Returns a Vec<u8>, with each byte written from the BinaryNumeralString
+    /// in little-endian bit order.
+    pub fn to_bytes_le(&self) -> Vec<u8> {
+        let mut data = Vec::with_capacity((self.0.len() + 7) / 8);
+        let mut acc = 0;
+        let mut shift = 0;
+        for n in &self.0 {
+            acc += n << shift;
+            shift += 1;
+            if shift == 8 {
+                data.push(acc);
+                acc = 0;
+                shift = 0;
+            }
+        }
+        data
+    }
+}
+
+impl NumeralString<PowerTwoRadix> for BinaryNumeralString {
+    fn is_valid(&self, radix: &PowerTwoRadix) -> bool {
+        self.0
+            .iter()
+            .fold(true, |acc, n| acc && radix.check_in_range(*n as u32))
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn split(&self, u: usize) -> (Self, Self) {
+        let mut front = self.0.clone();
+        let back = front.split_off(u);
+        (BinaryNumeralString(front), BinaryNumeralString(back))
+    }
+
+    fn concat(mut a: Self, mut b: Self) -> Self {
+        a.0.append(&mut b.0);
+        a
+    }
+
+    fn num_radix(&self, radix: &BigUint) -> BigUint {
+        let zero = BigUint::zero();
+        let one = BigUint::one();
+        // Check that radix == 2
+        assert!((radix & &one).is_zero());
+        assert_eq!(radix >> 1, one);
+        let mut res = zero;
+        for i in &self.0 {
+            res <<= 1;
+            if *i != 0 {
+                res += &one;
+            }
+        }
+        res
+    }
+
+    fn str_radix(mut x: BigUint, radix: &BigUint, m: usize) -> Self {
+        let one = BigUint::one();
+        // Check that radix == 2
+        assert!((radix & &one).is_zero());
+        assert_eq!(radix >> 1, one);
+        let mut res = vec![0; m];
+        for i in 0..m {
+            res[m - 1 - i] = if (&x & &one).is_zero() { 0 } else { 1 };
```

Comment:
This could be:
```
if x.is_odd() {
    res[m - 1 - i] = 1;
}
```
If we wanted to hide the bit magic behind the is_odd and take advantage of the fact that the vector is initialized with 0s.